### PR TITLE
Formalize the mkpool utility

### PR DESCRIPTION
**Pull request by stscieisenhamer**
**Description**

In investigating issues and requests from commissioning, the need for `mkpool` became apparent. However, the API had not been used nor modified since the conception of the association project.

The PR brings the code up-to-date with the current state of associations and implements the command-line tool `asn_make_pool`.

Checklist
- [x] Tests
- [x] Documentation
- [x] Change log
- [x] Milestone
- [x] Label(s)

### CODE REVIEW

File: scripts/asn_make_pool [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#!/user/bin/env python
```
Comment (tapastro) [inline scripts/asn_make_pool:1]: user -> usr

**Review by tapastro — CHANGES_REQUESTED** [review]
When I try to run `asn_make_pool pool='test.pool' data=*uncal.fits` in a directory with a few _uncal.fits files, I get

> usage: asn_make_pool [-h] [-v] pool data [data ...] asn-candidate dms-note is-imprt is-psf pntgtype
asn_make_pool: error: the following arguments are required: asn-candidate, dms-note, is-imprt, is-psf, pntgtype


Not clear that I'm using pool/data properly, but it appears that the rest of the flags are considered required input.

(If I don't use `pool=` and just try `asn_make_pool test.pool *uncal.fits`, which may appear obviously wrong already, the lack of parameter setting caused the script to try to use some of the uncal inputs as values for asn-candidate, dms-note, etc.)

**Comment by stscieisenhamer** [comment]
I definitely broke stuff with the command line. Fixed and added tests.

**Comment by tapastro** [comment]
Hmm, not sure if I'm using it inappropriately or if there's a bug - here's the traceback and the command that caused it:

`(jwstdev) wintermute:image tpauly$ asn_make_pool testpool.csv *uncal.fits
Traceback (most recent call last):
  File "/Users/tpauly/PycharmProjects/jwst/jwst/associations/pool.py", line 97, in write
    super(AssociationPool, self).write(
  File "/Users/tpauly/miniconda3/envs/jwstdev/lib/python3.9/site-packages/astropy/table/connect.py", line 129, in __call__
    self.registry.write(instance, *args, **kwargs)
  File "/Users/tpauly/miniconda3/envs/jwstdev/lib/python3.9/site-packages/astropy/io/registry/core.py", line 354, in write
    return writer(data, *args, **kwargs)
  File "/Users/tpauly/miniconda3/envs/jwstdev/lib/python3.9/site-packages/astropy/io/ascii/connect.py", line 26, in io_write
    return write(table, filename, **kwargs)
  File "/Users/tpauly/miniconda3/envs/jwstdev/lib/python3.9/site-packages/astropy/io/ascii/ui.py", line 837, in write
    writer.write(table, output)
  File "/Users/tpauly/miniconda3/envs/jwstdev/lib/python3.9/site-packages/astropy/io/ascii/fastbasic.py", line 167, in write
    self._write(table, output, {})
  File "/Users/tpauly/miniconda3/envs/jwstdev/lib/python3.9/site-packages/astropy/io/ascii/fastbasic.py", line 184, in _write
    writer.write(output, header_output, output_types)
  File "astropy/io/ascii/cparser.pyx", line 1125, in astropy.io.ascii.cparser.FastWriter.write
TypeError: unhashable type: 'numpy.ndarray'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/tpauly/miniconda3/envs/jwstdev/bin/asn_make_pool", line 7, in <module>
    exec(compile(f.read(), __file__, 'exec'))
  File "/Users/tpauly/PycharmProjects/jwst/scripts/asn_make_pool", line 12, in <module>
    pool.write(pool_file, overwrite=True)
  File "/Users/tpauly/PycharmProjects/jwst/jwst/associations/pool.py", line 105, in write
    super(AssociationPool, self).write(
  File "/Users/tpauly/miniconda3/envs/jwstdev/lib/python3.9/site-packages/astropy/table/connect.py", line 129, in __call__
    self.registry.write(instance, *args, **kwargs)
  File "/Users/tpauly/miniconda3/envs/jwstdev/lib/python3.9/site-packages/astropy/io/registry/core.py", line 354, in write
    return writer(data, *args, **kwargs)
  File "/Users/tpauly/miniconda3/envs/jwstdev/lib/python3.9/site-packages/astropy/io/ascii/connect.py", line 26, in io_write
    return write(table, filename, **kwargs)
  File "/Users/tpauly/miniconda3/envs/jwstdev/lib/python3.9/site-packages/astropy/io/ascii/ui.py", line 837, in write
    writer.write(table, output)
  File "/Users/tpauly/miniconda3/envs/jwstdev/lib/python3.9/site-packages/astropy/io/ascii/fastbasic.py", line 167, in write
    self._write(table, output, {})
  File "/Users/tpauly/miniconda3/envs/jwstdev/lib/python3.9/site-packages/astropy/io/ascii/fastbasic.py", line 184, in _write
    writer.write(output, header_output, output_types)
  File "astropy/io/ascii/cparser.pyx", line 1125, in astropy.io.ascii.cparser.FastWriter.write
TypeError: unhashable type: 'numpy.ndarray'
`

File: docs/jwst/associations/overview.rst [hunk 2]
```diff
@@ -126,13 +126,8 @@ as their respective DataModels.
...
-
 Utilities
 =========
 
-Other useful utilities for creating and manipulating associations:
-
-- :ref:`asn-from-list`
-- :ref:`asn-generate`
-- *many other TBD*
+There are a number of utilities to create user-specific associations which are
```
Comment (hbushouse) [inline docs/jwst/associations/overview.rst:132]: Either insert a comma after "assocations", or change "which" to "that"

File: jwst/associations/mkpool.py [hunk 3]
```diff
@@ -1,47 +1,223 @@
...
+    """Create an association pool from a list of FITS files.
+
+    Normally, association pools and the associations generated from those pools
+    are created by the automatic ground system process. Users should download
+    and modify those pools if need be. If desired, this function can be used to
+    create pools from scratch using a list of FITS files. Once created, the
+    :py:func:`~jwst.associations.generate` can be used to create associations
+    from these pools.
+
+    A number of columns used by the Association rules cannot be derived from the header
```
Comment (hbushouse) [inline jwst/associations/mkpool.py:45]: perhaps insert "pool" ahead of "columns", just to make it painfully obvious what columns we're talking about

File: jwst/associations/mkpool.py [hunk 4]
```diff
@@ -1,47 +1,223 @@
...
+        The specification can be either as a list of 2-tuples, as presented above, or
+        as a single string representation of the list. Using the previous example, the
+        following is also a valid input:
+        "[('c1000', 'background'), ('c1001', 'coronographic')]"
+    - dms_note
+          Notes from upstream processing of the downlinked data that may be pertinent
+          to the quality of the data. Currently the value "wfsc_los_jitter" is used
+          by the Level 2 wavefront sensing rule, Asn_Lv2WFSC, to ignore exposures.
+    - is_imprt
+          A 't' indicates the exposure is an imprint exposure.
```
Comment (hbushouse) [inline jwst/associations/mkpool.py:66]: Perhaps insert "NIRSpec" ahead of "imprint"

**Review by hbushouse — COMMENTED** [review]
Looks OK to my eyes. Does it work as intended?

**Comment by stscieisenhamer** [comment]
@tapastro are the data available to me to see what is going on?

File: docs/jwst/associations/overview.rst (hunk 2 above)
Comment (stscieisenhamer) [reply docs/jwst/associations/overview.rst:132]: ✅

File: jwst/associations/mkpool.py (hunk 3 above)
Comment (stscieisenhamer) [reply jwst/associations/mkpool.py:45]: ✅

File: jwst/associations/mkpool.py (hunk 4 above)
Comment (stscieisenhamer) [reply jwst/associations/mkpool.py:66]: ✅

**Comment by tapastro** [comment]
> @tapastro are the data available to me to see what is going on?

Almost certainly! `/Users/eisenham/test_bigdata/jwst-pipeline/dev/nircam/image/` is the equivalent of my working directory when I got that bug 🙂

**Comment by stscieisenhamer** [comment]
@tapastro ok, try it one more time...will get this right eventually...

**Comment by tapastro** [comment]
> @tapastro ok, try it one more time...will get this right eventually...

Hm, more interesting results this time! I can get single files to create a valid-looking association, but I get the same error when doing more than one - however, I do have one set of multiple files that did work, so something strange is going on.

**Comment by tapastro** [comment]
Utility is working for me, but the output pool file columns seem to come out in a random order, judging by repeated runs on the same input list. I suppose comparisons on a loaded pool may not care, but any text check or by-eye usage might suffer. Otherwise looks good.

**Review by tapastro — APPROVED** [review]

**Review by hbushouse — APPROVED** [review]

**Pull request merged**